src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/Target_X86/X86ReadyToRunHelperNode.cs

```diff
@@ -3,22 +3,75 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Collections.Generic;
-using System.Diagnostics;
+
 using ILCompiler.DependencyAnalysis.X86;
+
 using Internal.TypeSystem;
-using ILCompiler;
 
 namespace ILCompiler.DependencyAnalysis
 {
     /// <summary>
-    /// X64 specific portions of ReadyToRunHelperNode
+    /// X86 specific portions of ReadyToRunHelperNode
     /// </summary>
     public partial class ReadyToRunHelperNode
     {
         protected override void EmitCode(NodeFactory factory, ref X86Emitter encoder, bool relocsOnly)
         {
-            throw new NotImplementedException();
+            switch (Id)
+            {
+                case ReadyToRunHelperId.VirtualCall:
+                    {
+                        throw new NotImplementedException();
+                    }
+
+                case ReadyToRunHelperId.GetNonGCStaticBase:
+                    {
+                        MetadataType target = (MetadataType)Target;
+                        bool hasLazyStaticConstructor = factory.TypeSystemContext.HasLazyStaticConstructor(target);
+                        encoder.EmitMOV(encoder.TargetRegister.Result, factory.TypeNonGCStaticsSymbol(target));
+
+                        if (!hasLazyStaticConstructor)
+                        {
+                            encoder.EmitRET();
+                        }
+                        else
+                        {
+                            // We need to trigger the cctor before returning the base. It is stored at the beginning of the non-GC statics region.
+                            encoder.EmitMOV(encoder.TargetRegister.Arg0, factory.TypeNonGCStaticsSymbol(target), -NonGCStaticsNode.GetClassConstructorContextStorageSize(factory.Target, target));
+
+                            AddrMode initialized = new AddrMode(encoder.TargetRegister.Arg0, null, factory.Target.PointerSize, 0, AddrModeSize.Int32);
+                            encoder.EmitCMP(ref initialized, 1);
+                            encoder.EmitRETIfEqual();
+
+                            encoder.EmitMOV(encoder.TargetRegister.Arg1, encoder.TargetRegister.Result);
+                            encoder.EmitJMP(factory.HelperEntrypoint(HelperEntrypoint.EnsureClassConstructorRunAndReturnNonGCStaticBase));
+                        }
+                    }
+                    break;
+
+                case ReadyToRunHelperId.GetThreadStaticBase:
+                    {
+                        throw new NotImplementedException();
```

Comment:
I was also working on a version of this patch: https://github.com/am11/corert/commit/dfa63ecbeb776de4b33e8589b99e58ddcf728fc3, these cases were implemented using `EmitLEA` and `EmitLEAQ`. Could you squeeze those in? It compiles on Linux. :)

Reply:
`EmitLEAQ` will be replaced by `EmitMOV` on x86. I've added a commit that generates `int 3` instructions for these helpers - that way the compilation will proceed to the next problem and we'll get broken into the debugger when this is reached at runtime. I don't want to blindly add untested code because this is generally better to step through in a debugger to verify it really does the right thing. Porting these should be straightforward except for the delegate creation helper because we don't have enough registers to pass the 3 arguments to it. We can do something dirty when we get to that.

I've added a couple more things that should help Linux to successfully compile. It doesn't compile on Windows because CoreCLR uses a weird exception handling model on x86 Windows and RyuJIT is generating that (I believe it uses normal funclet-based model on x86 Unix). We'll need to fix up RyuJIT to use `FEATURE_EH_FUNCLETS` on Windows for CoreRT too. I don't think I'll get to that anytime soon. x86 Linux might be more realistic.